base/strings/util.jl

```diff
@@ -336,17 +342,22 @@ function chomp(s::AbstractString)
     (j < 1 || s[j] != '\r') && (return SubString(s, 1, j))
     return SubString(s, 1, prevind(s,j))
 end
-function chomp(s::String)
-    i = lastindex(s)
-    if i < 1 || codeunit(s,i) != 0x0a
-        return @inbounds SubString(s, 1, i)
-    elseif i < 2 || codeunit(s,i-1) != 0x0d
-        return @inbounds SubString(s, 1, prevind(s, i))
+
+@assume_effects :removable :foldable function chomp(s::Union{String, SubString{String}})
```

Comment:
@Keno how safe is marking consistency for string functions? (since they're in a weird semi-mutable place)

Reply:
Mutating a string is UB, so fine from that perspective I think

Reply:
in that, case I think the only issue is annotating stringer effects than should be necessary. is anything other than consistent, effect_free needed? the compiler should be able to figure out termination here.

Reply:
It doesn't, though.

Reply:
is the substring constructor not analyzed as terminating? if so, the fix should be there. in general it's a good idea to cover as little code as possible by effect annotations since that lowers the chance of lying to the compiler

Reply:
Good idea. But the substring constructor is a) generic over AbstractString, and b) only guaranteed to be nothrow if the indices are valid (as asserted by `@inbounds`). What's the best approach then? Make a wrapper method for `Union{String, SubString{String}}` with assumed effects?

Reply:
I think the substring method as is should be sufficient for `terminates_locally`. I'd be fine with annotating `nothrow` on this method. The main reason to narrow the effect annotations is that global termination is scary to annotate since if someone in the future adds a method that creates an infinite loop, the global termination annotation can make the debugging there a lot worse (since it can do things like corrupting stack traces)

Reply:
Ah, but then we don't gain much from annotating the substring constructor. Since then, `chomp` still doesn't get consistency, nothrow, effect_free, or terminates_globally, so it'd still need almost the same annotations.
We can also just remove the assume effect on `chomp` if it's not worth the hassle.

Reply:
I'm also fine with removing the effect annotation completely.

Reply:
eh, this is fine